### PR TITLE
Fixing genesis multisig

### PR DESCRIPTION
**Pull request by mrsmkl**
I guess the previous version of this PR was closed when `next` was merged into `master`.
The multisig actor was changed so that the signers now have to be resolved actor addresses. So here I add a new actor for each address that occurs in the multisigs.

Also changed checking the verified state so that it checks for the actor address instead of key address.

### CODE REVIEW

File: chain/gen/genesis/genesis.go [hunk 1]
```diff
@@ -332,8 +332,32 @@ func createAccount(ctx context.Context, bs bstore.Blockstore, cst cbor.IpldStore
 			return xerrors.Errorf("failed to create empty map: %v", err)
 		}
 
+		var signers []address.Address
+
+		for _, e := range ainfo.Signers {
+			idAddress, _ := keyIDs[e]
+			// Check if actor already exists
+			_, err := state.GetActor(e)
```
Comment (mrsmkl) [inline chain/gen/genesis/genesis.go:340]: Check if actor was already created.

File: chain/gen/genesis/genesis.go [hunk 2]
```diff
@@ -332,8 +332,32 @@ func createAccount(ctx context.Context, bs bstore.Blockstore, cst cbor.IpldStore
...
+			if err == nil {
+				continue
+			}
+			st, err := cst.Put(ctx, &account.State{Address: e})
+			if err != nil {
+				return err
+			}
+			err = state.SetActor(idAddress, &types.Actor{
```
Comment (mrsmkl) [inline chain/gen/genesis/genesis.go:348]: If an actor corresponding to ID doesn't exists, add an empty actor with zero balances.

File: chain/gen/genesis/t01_init.go [hunk 3]
```diff
@@ -71,6 +98,29 @@ func SetupInitActor(bs bstore.Blockstore, netname string, initialActors []genesi
 		if err := amap.Put(adt.AddrKey(ainfo.Owner), &value); err != nil {
 			return nil, nil, err
 		}
+	} else if rootVerifier.Type == genesis.TMultisig {
+		var ainfo genesis.MultisigMeta
+		if err := json.Unmarshal(rootVerifier.Meta, &ainfo); err != nil {
```
Comment (mrsmkl) [inline chain/gen/genesis/t01_init.go:103]: Same as above but for genesis multisig.

File: chain/gen/genesis/t01_init.go [hunk 4]
```diff
@@ -33,9 +33,35 @@ func SetupInitActor(bs bstore.Blockstore, netname string, initialActors []genesi
 	amap := adt.MakeEmptyMap(store)
 
 	keyToId := map[address.Address]address.Address{}
+	counter := int64(AccountStart)
```
Comment (mrsmkl) [inline chain/gen/genesis/t01_init.go:36]: Running counter for the IDs.

File: chain/gen/genesis/t01_init.go [hunk 5]
```diff
@@ -33,9 +33,35 @@ func SetupInitActor(bs bstore.Blockstore, netname string, initialActors []genesi
...
+			var ainfo genesis.MultisigMeta
+			if err := json.Unmarshal(a.Meta, &ainfo); err != nil {
+				return nil, nil, xerrors.Errorf("unmarshaling account meta: %w", err)
+			}
+			for _, e := range ainfo.Signers {
+
+				if _, ok := keyToId[e]; ok {
+					continue
+				}
+
```
Comment (mrsmkl) [inline chain/gen/genesis/t01_init.go:49]: Add all signers to lookup table.

File: chain/gen/genesis/t01_init.go [hunk 6]
```diff
@@ -48,12 +74,13 @@ func SetupInitActor(bs bstore.Blockstore, netname string, initialActors []genesi
 			return nil, nil, xerrors.Errorf("unmarshaling account meta: %w", err)
 		}
 
-		fmt.Printf("init set %s t0%d\n", ainfo.Owner, AccountStart+int64(i))
+		fmt.Printf("init set %s t0%d\n", ainfo.Owner, counter)
```
Comment (mrsmkl) [inline chain/gen/genesis/t01_init.go:77]: Perhaps also here we should check that the key isn't duplicate?

**Comment by magik6k** [comment]
Merged in https://github.com/filecoin-project/lotus/pull/3174

**Pull request closed without merging**